File: kotlin/src/main/kotlin/com/automation/agent/obs/Exporters.kt
/*
 * Span-exporter construction for the obs package.
 *
 * Builds the one exporter selected by config. The application names no vendor: every OTLP backend is
 * reached through [EXPORTER_OTLP] + endpoint, and [EXPORTER_GCP] is the one convenience path (Cloud
 * Trace via Application Default Credentials).
 */
package com.automation.agent.obs

import com.google.cloud.opentelemetry.trace.TraceExporter
import io.opentelemetry.exporter.logging.LoggingSpanExporter
import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter
import io.opentelemetry.sdk.trace.export.SpanExporter

/**
 * Build the span exporter for [cfg] exporter. The caller ([init]) has already rejected
 * [EXPORTER_NONE] (no exporter) and any unknown value, so this handles only the three real sinks.
 *
 * @throws IllegalArgumentException on an [EXPORTER_OTLP] config with no endpoint, or an unknown
 *   exporter.
 */
internal fun newExporter(cfg: Config): SpanExporter = when (cfg.exporter) {
    EXPORTER_CONSOLE -> LoggingSpanExporter.create()
    EXPORTER_OTLP -> {
        val endpoint = cfg.otlpEndpoint.trim()
        // config validates this, but guard so a direct caller fails loudly rather than silently
        // exporting nowhere.
        require(endpoint.isNotEmpty()) { "obs: exporter \"$EXPORTER_OTLP\" requires an OTLP endpoint" }
        val builder = OtlpHttpSpanExporter.builder().setEndpoint(endpoint)
        for ((key, value) in parseOtlpHeaders(cfg.otlpHeaders)) {
            builder.addHeader(key, value)
        }
        builder.build()
    }
    // No project id: the Cloud Trace exporter detects it from Application Default Credentials / the
    // metadata server, matching how the rest of the cloud path authenticates.
    EXPORTER_GCP -> TraceExporter.createWithDefaultConfiguration()
    else -> throw IllegalArgumentException("obs: unknown OTEL_TRACES_EXPORTER \"${cfg.exporter}\"")
}

/**
 * Parse the standard OTEL_EXPORTER_OTLP_HEADERS form — comma-separated key=value pairs (e.g.
 * "api-key=secret,env=prod") — into a header map. Blank entries and entries without a key are
 * skipped; only the first "=" splits, so a value may contain "=".
 */
fun parseOtlpHeaders(raw: String): Map<String, String> {
    val out = LinkedHashMap<String, String>()
    for (pairRaw in raw.split(",")) {
        val pair = pairRaw.trim()
        if (pair.isEmpty()) continue
        val eq = pair.indexOf('=')
        // No "=", or an empty key: skip rather than record a keyless header.
        if (eq <= 0) continue
        val key = pair.substring(0, eq).trim()
        if (key.isEmpty()) continue
        out[key] = pair.substring(eq + 1).trim()
    }
    return out
}

File: javascript/src/obs/exporters.ts
/**
 * Span-exporter construction for the obs package.
 *
 * Builds the one exporter selected by config. The application names no vendor: every OTLP backend
 * is reached through {@link TracesExporter.Otlp} + endpoint, and {@link TracesExporter.Gcp} is the
 * one convenience path (Cloud Trace via Application Default Credentials).
 */

import { ConsoleSpanExporter, type SpanExporter } from '@opentelemetry/sdk-trace-base';

import { type Config, TracesExporter } from './obs';

/**
 * Build the span exporter for `cfg.exporter`. The caller ({@link init}) has already rejected
 * {@link TracesExporter.None} (no exporter) and any unknown value, so this handles only the three
 * real sinks.
 *
 * The OTLP and Cloud Trace exporters are imported dynamically, not at module load: they pull in
 * heavy transitive dependencies, and the default no-exporter path must not pay that import cost.
 *
 * @throws Error on an {@link TracesExporter.Otlp} config with no endpoint, or an unknown exporter.
 */
export async function newExporter(cfg: Config): Promise<SpanExporter> {
  switch (cfg.exporter) {
    case TracesExporter.Console:
      return new ConsoleSpanExporter();
    case TracesExporter.Otlp: {
      const endpoint = (cfg.otlpEndpoint ?? '').trim();
      if (endpoint === '') {
        // config validates this, but guard so a direct caller fails loudly rather than silently
        // exporting nowhere.
        throw new Error(`obs: exporter ${JSON.stringify(TracesExporter.Otlp)} requires an OTLP endpoint`);
      }
      const { OTLPTraceExporter } = await import('@opentelemetry/exporter-trace-otlp-http');
      const headers = parseOtlpHeaders(cfg.otlpHeaders ?? '');
      return new OTLPTraceExporter({
        url: endpoint,
        headers: Object.keys(headers).length > 0 ? headers : undefined,
      });
    }
    case TracesExporter.Gcp: {
      // No project id: the Cloud Trace exporter detects it from Application Default Credentials /
      // the metadata server, matching how the rest of the cloud path authenticates.
      const { TraceExporter } = await import('@google-cloud/opentelemetry-cloud-trace-exporter');
      return new TraceExporter();
    }
    default:
      throw new Error(`obs: unknown OTEL_TRACES_EXPORTER ${JSON.stringify(cfg.exporter)}`);
  }
}

/**
 * Parse the standard `OTEL_EXPORTER_OTLP_HEADERS` form — comma-separated key=value pairs (e.g.
 * `"api-key=secret,env=prod"`) — into a header map. Blank entries and entries without a key are
 * skipped; only the first `=` splits, so a value may contain `=`.
 */
export function parseOtlpHeaders(raw: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const pairRaw of raw.split(',')) {
    const pair = pairRaw.trim();
    if (pair === '') {
      continue;
    }
    const eq = pair.indexOf('=');
    if (eq <= 0) {
      // No `=`, or an empty key: skip rather than record a keyless header.
      continue;
    }
    const key = pair.slice(0, eq).trim();
    if (key === '') {
      continue;
    }
    out[key] = pair.slice(eq + 1).trim();
  }
  return out;
}
